### PR TITLE
Install only new Odoo modules

### PR DESCRIPTION
**Pull request by enricostano**
Query to extract new or not installed modules from a list of modules:
```SQL
SELECT name FROM (
    VALUES ('base'), ('web')
) AS t (name)
EXCEPT
SELECT name FROM ir_module_module WHERE state = 'installed';
```

### CODE REVIEW

File: tasks/main.yml [hunk 1]
```diff
@@ -106,11 +106,32 @@
...
+  vars:
+    - joiner: "'), ('"
+  tags:
+    - mods
+
 - name: Init Odoo database
   become: yes
   become_user: "{{ odoo_role_odoo_user }}"
-  command: "{{ odoo_role_odoo_python_path }} {{ odoo_role_odoo_bin_path }} -c {{ odoo_role_odoo_config_path }}/odoo.conf -d {{ odoo_role_odoo_db_name }} --init {{ odoo_role_odoo_core_modules }} --stop-after-init --without-demo=all --logfile=/dev/stdout --log-level=warn"
+  command: "{{ odoo_role_odoo_python_path }} {{ odoo_role_odoo_bin_path }} -c {{ odoo_role_odoo_config_path }}/odoo.conf -d {{ odoo_role_odoo_db_name }} --init {{ modules_to_install }} --stop-after-init --without-demo=all --logfile=/dev/stdout --log-level=warn"
```
Comment (raneq) [inline tasks/main.yml]:
```suggestion
  command: "{{ odoo_role_odoo_python_path }} {{ odoo_role_odoo_bin_path }} -c {{ odoo_role_odoo_config_path }}/odoo.conf -d {{ odoo_role_odoo_db_name }} --init {{ modules_to_install.stdout }} --stop-after-init --without-demo=all --logfile=/dev/stdout --log-level=warn"
```

File: tasks/main.yml [hunk 2]
```diff
@@ -106,11 +106,32 @@
...
+    FROM (VALUES ('{{ odoo_role_odoo_core_modules.split(',') | join(joiner) }}'))
+    AS t (name)
+    EXCEPT
+    SELECT name
+    FROM ir_module_module
+    WHERE state = 'installed';"
+  register: modules_to_install
+  changed_when: false
+  vars:
+    - joiner: "'), ('"
```
Comment (enricostano) [inline tasks/main.yml]: `separator`

File: tasks/main.yml [hunk 3]
```diff
@@ -106,11 +106,32 @@
     group: "{{  odoo_role_odoo_group }}"
   notify: restart odoo
 
+- name: Build the list of new modules to install
+  become: yes
+  become_user: "{{ odoo_role_odoo_user }}"
+  shell: >
+    psql {{ odoo_role_odoo_db_name }} -tAc
+    "SELECT name
+    FROM (VALUES ('{{ odoo_role_odoo_core_modules.split(',') | join(joiner) }}'))
```
Comment (raneq) [inline tasks/main.yml]: `join(joiner)` doesn't tell much, can you think about other name? :)

File: tasks/community-modules.yml [hunk 4]
```diff
@@ -5,19 +5,12 @@
...
 
 - name: Deploy community roles with pip
   pip:
     requirements: "{{ odoo_role_odoo_modules_path }}/requirements.txt"
     virtualenv: "{{ odoo_role_odoo_venv_path }}"
-
-- name: Update Odoo modules list
-  become: yes
-  become_user: "{{ odoo_role_odoo_user }}"
-  command: "{{ odoo_role_odoo_python_path }} {{ odoo_role_odoo_bin_path }} -c {{ odoo_role_odoo_config_path }}/odoo.conf -d {{ odoo_role_odoo_db_name }} --update all --stop-after-init --without-demo=all --logfile=/dev/stdout --log-level=warn"
```
Comment (enricostano) [inline tasks/community-modules.yml]: It seems that it's not needed [anymore](https://github.com/odoo/odoo/issues/11802#issuecomment-387314916)

File: tasks/main.yml [hunk 5]
```diff
@@ -106,12 +106,75 @@
...
+    --log-level=warn
+  when: db_initialized.stdout == "0"
   notify: restart odoo
+  tags:
+    - mods
 
 - import_tasks: community-modules.yml
 
+- debug:
+    msg: "{{ odoo_role_community_odoo_modules }}"
```
Comment (enricostano) [inline tasks/main.yml]: I don't know why... but this variable doesn't get update when we modify it in `group_vars/all.yml` :rage4:

Comment (enricostano) [reply tasks/main.yml]: Solved by https://gitlab.com/coopdevs/odoo-test-provisioning/merge_requests/2/diffs?commit_id=1f714f99ba57cbb17e8e97a1ee6048fdf0de2943

File: tasks/main.yml (hunk 3 above)
Comment (enricostano) [reply tasks/main.yml]: it's called `separator` now

**Comment by enricostano** [comment]
@raneq @danypr92 it's not WIP anymore, would you mind to review again and possibly test it? Thanks!

**Comment by raneq** [comment]
@enricostano I've tested it and gives me some error. See:

```
TASK [coopdevs.odoo-role : Build the list of new modules to install] **********************************************************************************************************************************************
task path: /home/raneq/coopdevs/odoo/odoo-role/tasks/main.yml:136
 [WARNING]: Using world-readable permissions for temporary files Ansible needs to create when becoming an unprivileged user. This may be insecure. For information on securing this, see
https://docs.ansible.com/ansible/become.html#becoming-an-unprivileged-user

changed: [odoo-test.local] => {
  "changed": true,
  "cmd":
  "psql odoo -tAc \"SELECT name FROM
    (VALUES ('sale_management'),
    ('account'), ('account_invoicing'), ('account_asset'),
    ('account_analytic_default'), ('account_bank_statement_import'),
    ('account_cancel'), ('analytic'), ('auth_crypt'), ('auth_signup'),
    ('base'), ('base_automation'), ('base_iban'), ('base_import'),
    ('base_setup'), ('base_vat'), ('base_vat_autocomplete'),
    ('contacts'), ('decimal_precision'), ('document'), ('product'),
    ('web'), ('web_diagram'), ('web_kanban_gauge'), ('web_planner'),
    ('web_settings_dashboard'), ('l10n_be'), ('account_banking_mandate'),('account_banking_pain_base'), ('account_banking_sepa_credit_transfer'), ('account_banking_sepa_direct_debit'), ('account_due_list'), ('account_financial_report'), ('account_payment_order'), ('account_payment_partner'), ('base_bank_from_iban'), ('base_technical_features'), ('contract'), ('contract_sale_invoicing'), ('contract_variable_quantity'), ('web_favicon'), ('web_no_bubble'), ('web_responsive'), ('web_searchbar_full_width'), ('mis_builder_budget')) AS t (name) EXCEPT SELECT name FROM ir_module_module WHERE state = 'installed';\"", "delta": "0:00:00.051544", "end": "2019-06-11 13:39:29.750333", "rc": 0, "start": "2019-06-11 13:39:29.698789", "stderr": "", "stderr_lines": [], "stdout": "mis_builder_budget\nweb_favicon", "stdout_lines": ["mis_builder_budget", "web_favicon"]}

TASK [coopdevs.odoo-role : Install only new Odoo modules] *********************************************************************************************************************************************************
task path: /home/raneq/coopdevs/odoo/odoo-role/tasks/main.yml:152
 [WARNING]: Using world-readable permissions for temporary files Ansible needs to create when becoming an unprivileged user. This may be insecure. For information on securing this, see
https://docs.ansible.com/ansible/become.html#becoming-an-unprivileged-user

NOTIFIED HANDLER coopdevs.odoo-role : restart odoo for odoo-test.local
changed: [odoo-test.local] => {
  "changed": true,
  "cmd": [
    "/opt/.odoo_venv/bin/python",
    "/opt/odoo/odoo-bin",
    "-c", "/etc/odoo/odoo.conf",
    "-d", "odoo",
    "--init", "mis_builder_budget,web_favicon",
    "--stop-after-init", "--without-demo=all",
    "--logfile=/dev/stdout", "--log-level=warn"
  ],
  "delta": "0:00:03.337316", "end": "2019-06-11 13:39:33.545689",
  "rc": 0,
  "start": "2019-06-11 13:39:30.208373",
  "stderr": "ERROR: couldn't create the logfile directory. Logging to the standard output.\n2019-06-11 13:39:31,872 9164 WARNING odoo odoo.modules.loading: invalid module names, ignored: web_favicon, mis_builder_budget", "stderr_lines": ["ERROR: couldn't create the logfile directory. Logging to the standard output.", "2019-06-11 13:39:31,872 9164 WARNING odoo odoo.modules.loading: invalid module names, ignored: web_favicon, mis_builder_budget"], "stdout": "", "stdout_lines": []}

```

**Comment by raneq** [comment]
My local odoo now is broken, but I may have broken it by cancelling a provisioning. I'll try later with a new instance.

**Comment by raneq** [comment]
Hi, fresh instance here.  It works!

The provisioning installed well and I check that the issue with names was due to my inventory, so that files/requirements.txt was missing modules from the variables, implying that we tried to make odoo "install" things that we didn't put there for it in the first place. Thus the `    "2019-06-11 13:39:31,872 9164 WARNING odoo odoo.modules.loading: invalid module names, ignored: web_favicon, mis_builder_budget"`

When, after installing, I modified both requirements and module list, it installed the missing module(s). So green light for me! If @danypr92 doesn't have any comment, I propose to merge this :)

**Pull request merged**